piet-gpu-hal/examples/shader/stack.comp

```diff
@@ -0,0 +1,217 @@
+    uint npush = inp;
+    uint stack = el_ix;
+    sh_npop[th_ix] = npop;
+    sh_npush[th_ix] = npush;
+    sh_stack[th_ix] = stack;
+    bool pending = true;
+
+    // Reduction tree for computation of monoid for this partition
+    for (uint j = 0; j < LG_PART_SIZE; j++) {
+        uint bit = 1 << j;
+        uint first = th_ix & (-2u * bit);
+        uint mid = first + bit - 1;
+        uint last = mid + bit;
+        // Combine monoid from first..first + bit and first + bit..first + bit * 2
+        // to first..first + bit * 2
+        barrier();
+        // In this section, only read from shared memory
+        uint push_mid = sh_npush[mid];
+        if ((th_ix & bit) != 0) {
+            uint m = max(push_mid, npop);
+            npush = m - npop + npush;
+            uint prepop = npop - local_pop;
+            npop = sh_npop[mid] + m - push_mid;
+            if (pending && push_mid > prepop) {
+                // Generate final output when it can be resolved in this partition
+                data[el_ix] = sh_stack[mid - prepop];
+                pending = false;
+            }
+        }
+        uint pop_last = sh_npop[last];
+        uint m = max(push_mid, pop_last);
+        uint push_last = sh_npush[last];
+        if (last - th_ix >= push_last) {
+            uint iprev = last - th_ix - push_last;
+            if (iprev < m - pop_last) {
+                stack = sh_stack[mid - pop_last - iprev];
+            } else {
+                // TODO: not needed
+                stack = 99;
+            }
+        }
+        barrier();
+        // Write local values to shared memory
+        sh_npop[th_ix] = npop;
+        sh_npush[th_ix] = npush;
+        sh_stack[th_ix] = stack;
+    }
+
+    // If pending, then stack depth of result of stack snapshot immediately
+    // preceding this partition.
+    uint prepop = npop - local_pop;
+
+    barrier();
+    // Publish local aggregate
+    uint part_npush = sh_npush[PART_SIZE - 1];
+    uint part_npop = sh_npop[PART_SIZE - 1];
+    if (PART_SIZE - 1 - th_ix < part_npush) {
+        scratch[el_ix] = stack;
+    }
+    uint state_base = part_ix * STATE_SIZE;
+    if (th_ix == PART_SIZE - 1) {
+        state[state_base + STATE_NPOP_LOCAL] = npop;
+        state[state_base + STATE_NPUSH_LOCAL] = npush;
+        uint flags = FLAG_LOCAL;
+        if (part_ix == 0) {
+            flags = FLAG_DONE;
+            state[state_base + STATE_NPUSH_PREFIX] = npush;
+        }
+        state[state_base + STATE_FLAGS] = part_ix == 0 ? FLAG_DONE : FLAG_LOCAL;
+    }
+
+    // Do lookback
+    // Note: unlike prefix sum example, we need all threads active to do stack reduction.
+    if (part_ix != 0) {
+        bool backlink = false;
+        uint look_back_ix = part_ix - 1;
+        while (true) {
+            if (th_ix == PART_SIZE - 1) {
+                // This is for evaluation and will be removed or conditional
+                //atomicAdd(state[part_ix * STATE_SIZE + 5], 1);
```

Comment:
Remove or expand comment. STATE_SIZE depends on this as well, right? If so, it too should be conditional.